### ISSUE TITLE
Add event listeners

### ISSUE THREAD

**Issue by sjelfull** [sjelfull/craft-supersearch#8]
- [x] When elements is saved
- [x] When elements is deleted
- [x] When structure elements is moved
- [ ] ?

**Comment by timkelty** [comment]
Handling related elements can be tricky here.

E.g. When you save a category, you may need to re-index every element that has a relationship to that category.

**Comment by sjelfull** [comment]
Yeah, I've hit that wall before. I haven't decided how to tackle that, but have some ideas.

Think the easiest is to
- find all elements with a relationship to the element being updated
- repeat step 1 for all found elements
- de-duplicate result
- re-index resulting elements

**Comment by timkelty** [comment]
Yep, that's how I've handled it too.

As these things grow quickly/exponentially, I've found it crucial that the mapping part be task-based and not execute directly on save. Even executing the sprawling relationship queries on-save has bogged down the CP for me, so on save I fire a task that first collects all the related elements, then maps them, then batches them to Algolia.

Some of this is similar to how Search Plus for Craft 2 handles syncing: https://transition.topshelfcraft.com/software/craft/searchplus
